components/raftstore/src/store/fsm/peer.rs

```diff
@@ -729,9 +729,12 @@ where
     #[inline]
     fn on_loop_finished(&mut self) {
         let ready_concurrency = self.ctx.cfg.cmd_batch_concurrent_ready_max_count;
+        // allow propose pending commands iff all ongoing commands are
+        // persisted or committed.
         let should_propose = self.ctx.sync_write_worker.is_some()
             || ready_concurrency == 0
-            || self.fsm.peer.unpersisted_ready_len() < ready_concurrency;
+            || self.fsm.peer.unpersisted_ready_len() < ready_concurrency
+            || !self.fsm.peer.has_uncommitted_log();
```

Comment:
I suspect if it would cause bubble in the write flow pipeline. If propose early, the raft message can be sent out in this round. How about allowing propose when last - commit < threshold?